### PR TITLE
[FSDP2] Added backward prefetching

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119378
* #119302
* #118298
* #118755
* #118223
* #118136
* __->__ #118118
* #118017

This PR adds explicit backward prefetching to overlap communication and computation in backward (namely, needed for `reshard_after_forward=True` or `reshard_after_forward: int`). We do this by recording the post-forward order and using its reverse to approximate the backward order.

This works for the typical 1 forward / 1 backward training. However, for more complex schedules, this can run into some gaps:
- We need to know the _true end of backward_.
    - At the true of end of backward, we can clear our recorded post-forward order and pre-backward hook state, and we should wait on gradient reductions.
    - There is no easy way to know whether the current backward marks the true end of backward. Therefore, we introduce an API for the user to set this: `fsdp_module.set_is_last_backward(bool)`. For example, for pipeline parallelism's DFS cooldown backward, we can call `fsdp_module.set_is_last_backward(is_last_microbatch)`.
- When the user runs backward through only part of the model, our reverse-post-forward-order heuristic risks _mistargeted prefetches_ for unused modules, which would mean the module's parameters are all-gathered and not freed until the end of backward.
    - To error on the side of less memory usage (but no overlap), this PR introduces logic to check whether a module will need its unshard in the current backward (by recording the module's `forward` outputs' `grad_fn`s and querying the autograd engine).
    - Note that there may be _no_ overlap in backward for some parts due to no prefetching.
    - Note further that when running multiple backwards, if the user does not use `set_is_last_backward`, we may not be able to provide a meaningful error message, as the pre-backward hook could be erroneously cleared on the 1st backward.
    - In the future, we may expose more APIs from the autograd engine (similar to `_current_graph_task_execution_order`) to make the prefetching exact. (Currently, `_current_graph_task_execution_order` requires the `with torch.autograd.set_multithreading_enabled(False)`, which is too hard of a constraint as we cannot easily modify users' training loops. We can replace the multi-threading check with a device check. Moreover, in the partial backward case in this PR's unit test, I still hit an [internal assertion](https://github.com/pytorch/pytorch/blob/b816760a2f27adafb0b1dac4c032a2e97c690b29/torch/csrc/autograd/engine.cpp#L476), so some follow-up is required.)

<details>
<summary> Old Discussion </summary>

For discussion:
- The PR includes a counter `expected_backward_unshard_count` to mitigate mistargeted prefetches in backward. However, it can be seen as a necessary but not sufficient solution.
    - If a module's outputs do not require gradient, then we certainly do not need to unshard the module in backward.
    - However, if a module's outputs do require gradient, then we still may not need to unshard the module for _this_ backward (e.g. if the module did not contribute to `loss` for the current `loss.backward()`).
    - This counter will only address the first case but not the second. If we want to address the second, then we may need more info from the autograd engine.
- For now, I did not include any unit test to cover these behaviors, as I do not have a good example yet.
</details>

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: test/distributed/_composable/fsdp/_test_fully_shard_common.py [hunk 1]
```diff
@@ -53,3 +55,23 @@ def patch_reduce_scatter(new_reduce_scatter_tensor: Callable):
         yield
     finally:
         dist.reduce_scatter_tensor = orig_reduce_scatter
+
+
+@contextlib.contextmanager
+def patch_unshard(new_unshard: Callable):
+    orig_unshard = FSDPParamGroup.unshard
+    FSDPParamGroup.unshard = new_unshard
```
Comment (awgu) [inline test/distributed/_composable/fsdp/_test_fully_shard_common.py:63]: I do not have a better way of doing this patching context manager because of this assignment into the original method. If anyone has ideas, I can try.

Otherwise for now, we have boilerplate for every patch context.

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 2]
```diff
@@ -52,6 +59,11 @@ def __init__(
         self.all_gather_stream: torch.cuda.Stream = default_stream
         self.reduce_scatter_stream: torch.cuda.Stream = default_stream
         self.all_gather_state = AllGatherStateHolder()
+        self.post_forward_order: List[FSDPParamGroup] = []
+        self._post_forward_indices: List[int] = []
+        # Used to avoid mistargeted backward prefetches in the case that some
+        # module is used in forward but not in backward
+        self.expected_backward_unshard_count: int = 0
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:66]: maybe change ``expected_`` to ``estimated_`` before addressing the case of ``requires_grad=True`` but not part of ``loss.backward`` ?

[optional] I also spend some time to realize it's meant for remaining unfinished unshards since we are ``+/- = 1``. so I was thinking of ``estimated_unfinished_backward_unshard``

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 3]
```diff
@@ -177,14 +189,24 @@ def pre_forward(
     def post_forward(self, module: nn.Module, input: Any, output: Any):
         with torch.profiler.record_function("FSDP::post_forward"):
             self.reshard()
+            self._record_post_forward()
             self._training_state = TrainingState.IDLE
             return output
 
+    def _record_post_forward(self) -> None:
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:196]: seems we are merging ``_ExecOrderData`` into ``FSDPParamGroup``. where would we place ``_backward_prefetch_limit`` / ``_forward_prefetch_limit``? I remember you mentioned rate limiter will be cuda stream based instead of cpu based?

**Review by weifengpy — APPROVED** [review]

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 4]
```diff
@@ -262,6 +285,25 @@ def finalize_backward(self):
             torch.cuda.current_stream().wait_event(self._reduce_scatter_view_out_event)
             self._reduce_scatter_view_out_event = None
         self._training_state = TrainingState.IDLE
+        self._post_forward_indices.clear()
+        self.expected_backward_unshard_count = 0
+
+    def _prefetch_unshard(self):
+        if self._training_state == TrainingState.PRE_BACKWARD:
+            if not self._post_forward_indices:
+                msg = "Unexpected backward prefetching without running forward"
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:294]: Add `_module_fqn` into `msg`

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 5]
```diff
@@ -254,34 +276,75 @@ def _post_backward(self, *unused: Any):
...
+            if not self._post_forward_indices:
+                # Can happen if running backward >1 time through this module
+                # (with `retain_graph=True`)
+                return
+            curr_index = self._post_forward_indices.pop()
+            if (target_index := curr_index - 1) < 0:
+                return
+            target_fsdp_param_group = self.comm_ctx.post_forward_order[target_index]
+            if any(
+                torch._C._will_engine_execute_node(grad_fn)  # type: ignore[attr-defined]
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:297]: `_will_engine_execute_node` is `O(1)`, so this check is linear in the number of forward outputs (that require gradient) that the target parameter group was involved in. This seems tolerable to me.

File: torch/distributed/_composable/fsdp/fully_shard.py [hunk 6]
```diff
@@ -144,7 +144,20 @@ def reshard(self) -> None:
         Reshards the module's parameters by freeing unsharded parameters if
         needed. This method is *not* recursive.
         """
-        if (state := _get_module_fsdp_state(cast(nn.Module, self))) is None:
-            raise AssertionError(f"No FSDP state found on {self}")
+        state = self._get_fsdp_state()
         if fsdp_param_group := state._fsdp_param_group:
             fsdp_param_group.reshard()
+
+    def set_is_last_backward(self, is_last_backward: bool) -> None:
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/fully_shard.py:151]: We need this API (or similar) anyway for pipeline parallel DFS cooldown backward where there are back-to-back microbatch backwards.

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 7]
```diff
@@ -92,6 +96,13 @@ def __init__(
 
         # - Communication and communication/computation overlap
         self.comm_ctx = FSDPCommContext()
+        # Group's indices in the shared post-forward order
+        self._post_forward_indices: List[int] = []
+        # Used to avoid mistargeted backward prefetches when the module is used
+        # in forward but not in backward: for each forward, we record a tuple
+        # of the output's grad fns and later query the autograd engine whether
+        # any grad fn will execute in the current backward to know to prefetch.
+        self.all_forward_grad_fns: Set[Tuple[Any, ...]] = set()
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:105]:
```suggestion
        self.all_forward_grad_fns: Set[Tuple[torch.autograd.graph.Node, ...]] = set()
```

File: torch/distributed/_composable/fsdp/_fsdp_state.py [hunk 8]
```diff
@@ -158,16 +162,23 @@ def _post_forward(self, module: nn.Module, input: Any, output: Any) -> Any:
                 self._comm_ctx.all_gather_state = None  # free the all-gather result
             self._state_ctx.iter_forward_root = None
 
-    def _pre_backward(self, *unused: Any) -> None:
+    def _pre_backward(self, forward_grad_fns: Tuple[Any], *unused: Any) -> None:
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_state.py:165]:
```suggestion
    def _pre_backward(self, forward_grad_fns: Tuple[torch.autograd.graph.Node, ...], *unused: Any) -> None:
```

**Review by awgu — COMMENTED** [review]
~~To-do: see if we can make the type annotation more specific~~
updated

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 9]
```diff
@@ -216,16 +228,27 @@ def pre_forward(
...
+        post_forward_index = len(self.comm_ctx.post_forward_order)
+        self.comm_ctx.post_forward_order.append(self)
+        self._post_forward_indices.append(post_forward_index)
+
+    def pre_backward(self, forward_grad_fns: Tuple[Any, ...], *unused: Any):
         with torch.profiler.record_function("FSDP::pre_backward"):
             self._training_state = TrainingState.PRE_BACKWARD
             self.unshard()  # no-op if prefetched
             self.wait_for_unshard()
+            # Can be already removed if running backward >1 time
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:247]: "running backward >1 time": is this still referring to the current loop (not accumulating grad) ?

Comment (Dalian991) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:247]: @awgu

File: torch/distributed/_composable/fsdp/fully_shard.py (hunk 6 above)
Comment (weifengpy) [reply torch/distributed/_composable/fsdp/fully_shard.py:151]: is this mostly for PP? for FSDP-only, we would not need this? this is base on my understanding of final post-backward callback of flat-param FSDP

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py (hunk 9 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:247]: Sorry. This is referring to calling `.backward()` multiple times, where all but the last time would need to call with `retain_graph=True`.

Within a single `.backward()` call, I do not think we can run this multiple times.

File: torch/distributed/_composable/fsdp/fully_shard.py (hunk 6 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/fully_shard.py:151]: We need it any time we are running multiple backwards without a forward. This happens to appear in PP, but for some multi-modal models, I think it could happen with FSDP-only.

File: test/distributed/_composable/fsdp/test_fully_shard_comm.py [hunk 10]
```diff
@@ -94,5 +104,240 @@ def reduce_scatter_with_count(*args, **kwargs):
...
+        model, optim, inp = self._init_transformer(n_layers, reshard_after_forward)
+        events: List[EventType] = []
+        unshard_with_record = self._get_unshard_with_record(
+            FSDPParamGroup.unshard, events
+        )
+        post_backward_with_record = self._get_post_backward_with_record(
+            FSDPParamGroup.post_backward, events
+        )
+        # Check the order for normal 1 forward, 1 backward, 1 optimizer step
+        with patch_unshard(unshard_with_record), patch_post_backward(
```
Comment (wanchaol) [inline test/distributed/_composable/fsdp/test_fully_shard_comm.py:137]: Hmmm, it's a bit hard to follow the user workflow when using reshard_after_forward and backward prefetch. and the reshard_after_forward feature seems getting FSDP more complicated than my original thought..

**Review by wanchaol — COMMENTED** [review]
Could you add an user UX example about how they would use reshard_after_forward + backward prefetching?

File: test/distributed/_composable/fsdp/test_fully_shard_comm.py (hunk 10 above)
Comment (wanchaol) [reply test/distributed/_composable/fsdp/test_fully_shard_comm.py:137]: Oh I see, I thought there needs some special API for user to derive the backward order events for backward prefetching in the case of reshard_after_forward, if this happens automatically then it should be fine.

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 11]
```diff
@@ -254,34 +277,74 @@ def _post_backward(self, *unused: Any):
...
+        self._post_forward_indices.clear()
+        self.all_forward_grad_fns.clear()
+
+    def _prefetch_unshard(self):
+        if self._training_state == TrainingState.PRE_BACKWARD:
+            if not self._post_forward_indices:
+                # Can be cleared if running multiple `backward`s
+                return
+            curr_index = self._post_forward_indices.pop()
+            if (target_index := curr_index - 1) < 0:
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:293]: This means the root state where it does not have reshard_after_forward, so we don't need the prefetch unshard?

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 12]
```diff
@@ -254,34 +277,74 @@ def _post_backward(self, *unused: Any):
...
+                torch._C._will_engine_execute_node(grad_fn)  # type: ignore[attr-defined]
+                for forward_grad_fns in target_fsdp_param_group.all_forward_grad_fns
+                for grad_fn in forward_grad_fns
+            ):
+                with torch.profiler.record_function(
+                    "FSDP::backward_prefetch"
+                ), target_fsdp_param_group.use_training_state(
+                    TrainingState.PRE_BACKWARD
+                ):
+                    target_fsdp_param_group.unshard()
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:306]: hmmm I wonder if `target_fsdp_param_group` is always `self`? (I guess not). If it's not always self, it feels a bit weird to me that the current fsdp_param_group is actually prefetching another "target" fsdp_param_group

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 13]
```diff
@@ -92,6 +97,13 @@ def __init__(
 
         # - Communication and communication/computation overlap
         self.comm_ctx = FSDPCommContext()
+        # Group's indices in the shared post-forward order
+        self._post_forward_indices: List[int] = []
+        # Used to avoid mistargeted backward prefetches when the module is used
+        # in forward but not in backward: for each forward, we record a tuple
+        # of the output's grad fns and later query the autograd engine whether
+        # any grad fn will execute in the current backward to know to prefetch.
+        self.all_forward_grad_fns: Set[Tuple[Node, ...]] = set()
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:106]: nit: make it clear it's output grad fns, `all_forward_output_grad_fns`

**Review by wanchaol — APPROVED** [review]
looks good!

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py (hunk 11 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:293]: Yep! (In general, it is probably good to have the out-of-bounds index check, but yes, in this case, it would be the root.)

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py (hunk 12 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:306]: I think it is never `self`. I wonder why it is weird haha -- the currently executing group has to prefetch on behalf of the next group since the next group did not run yet.

Comment (wanchaol) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:306]: Oh it's more like OO design issue, intuitively I feel prefetching should happen in a higher level abstraction that manages ParamGroups instead of ParamGroup directly, a ParamGroup instance should morally should only care about its own state (its own unshard/shard/reshard), instead of other param groups.

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:306]: Ah yes that makes sense. There is definitely some leaky abstraction here (as evidenced even by the circular references). I did not have a solution to reorganize though 🤔

Comment (wanchaol) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:306]: I think it's fine to land as is, this is something we could think about after the initial PRs landed

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 14]
```diff
@@ -139,6 +139,10 @@ def foreach_reduce_scatter(
         # Record to mark the end of the reduce-scatter copy-in in the RS stream
         copy_in_event = torch.cuda.Event()
         copy_in_event.record()
+        # Only after the copy-in finishes can we free the gradients, which were
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py:142]: FYI: I am sneaking this change into this PR for convenience. I ran the stream-related unit tests, and it does not affect correctness -- we do not need to be outside the stream context to have a stream wait for an event, so it does not matter that we move this up. The motivation is that we can clear the unsharded gradients _before_ we allocate the reduce-scatter output since we are only copying into the reduce-scatter input. This lowers peak memory by a small amount.
